internal/cmd/msys2.go

```diff
@@ -0,0 +1,348 @@
+		return m
+	}
+
+	if _, err := exec.LookPath("cygpath.exe"); err != nil {
+		// This is an MSYS2 environment...without cygpath. We require cygpath
+		// for some initial setup.
+		return m
+	}
+
+	// At this point we have a valid MSYS2 environment.
+
+	// We set this to true first, to enable some of the other msys helper
+	// functions.
+	m.isMsys2Env = true
+
+	if v, err := m.toWinMixedPath("/"); err == nil {
+		m.winMixedRoot = v
+	}
+
+	if v, err := m.toWinMixedPath("/tmp"); err == nil {
+		m.winMixedTmp = v
+	}
+
+	return m
+}
+
+// Msys2 is the MSYS2 environment instance.
+var Msys2 = newMsys2()
+
+// isMsys2 returns true if the environment is an MSYS2 environment on Windows.
+func (m msys2) isMsys2() bool {
+	return m.isMsys2Env
+}
+
+// isIgnoredEnvKey returns true if the given key k must be ignored by our
+// conversion step.
+//
+// This tries to ensure that direnv.exe produces similar results than the MSYS2
+// env command.
+func (m msys2) isIgnoredEnvKey(k string) bool {
+
+	switch strings.ToUpper(k) {
+
+	case "":
+		return true
+
+	// If this is ignored, the "space dir" bash test fails with an error: rm:
+	// command not found
+	case "PATH":
+		return false
+
+	// If this is ignored, the "dump" bash test fails.
+	// Alternative solution: make the bash test msys2 aware.
+	case "LESSOPEN":
+		return false
+
+	// If this is ignored, the "python-layout" and "python-custom-virtual-env"
+	// bash test fails.
+	// Alternative solution: make the bash test msys2 aware.
+	case "VIRTUAL_ENV":
+		return false
+
+	// All remaining keys will be ignored, i.e. kept as received by MSYS2.
+	default:
+		return true
+	}
+}
+
+// isIgnoredEnvVal returns true if the given value v must be ignored by our
+// conversion step.
+//
+// This tries to ensure that direnv.exe produces similar results than the MSYS2
+// env command.
+func (m msys2) isIgnoredEnvVal(v string) bool {
+	if !m.winPathRegex.MatchString(v) {
+		// This does not look like a Windows path, so let's ignore it.
+		return true
+	}
+
+	// winPathRegex will match this Unix path: /mingw64/lib/pkgconfig:/mingw64/share/pkgconfig
+	// So let's double check.
+
+	// This looks like a Windows path, but is it really?
+	if m.winPathRegex2.MatchString(v) {
+		// Nope, this was a false positive, and is probably a Unix path, so let's ignore it.
+		return true
+	}
+
+	// Our best guess is that this is a Windows path, so don't ignore it.
+	return false
+}
+
+// execCygpath calls cygpath.exe and returns the output.
+func (m msys2) execCygpath(args ...string) (string, error) {
+	if !m.isMsys2Env {
+		return "", errMsys2Env
+	}
+
+	output, err := exec.Command("cygpath.exe", args...).Output()
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(string(output)), nil
+}
+
+// toWinMixedPath uses cygpath.exe to convert the Unix input path to a Windows
+// Mixed path.
+func (m msys2) toWinMixedPath(input string) (string, error) {
+	return m.execCygpath("-m", input)
```

Comment:
How much logic is it to port over if we wanted all the logic embedded in direnv?

I'm asking because Windows is a bit slow in spawning new programs.

Reply:
To implement cygpath, you would at minimum need to be able to parse /etc/fstab, which controls the drive mapping prefix for both Cygwin and msys (unless msys has an API for this that you can call?).  See also [the source code](https://cygwin.com/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=winsup/utils/cygpath.cc) for the bit that doesn't involve /etc/fstab.

(Oh, and in order to *find*/etc/fstab you'll probably need to run cygpath first, as there are many possible install locations for msys and cygwin; my msys is in a subdirectory of `~/scoop/apps/`, for example.)

Reply:
@zimbatm I agree, the spawning part can be slow. This PR already does some effort to reduce the amount of calls to cygpath. This specific function, for example, is called only during construction of the `msys2` object (which happens once during a single execution of direnv).

But you are right, it can be improved by implementing the relevant parts of cygpath itself, directly in Go. If we don't need a fully equivalent cygpath implementation, then we might get away with a heuristic based solution. Here is an example: since the current implementation only cares to know where `/` and `/tmp` are mapped to, we could potentially resolve those paths without using cygpath, by looking at the `TMP`, `TMPDIR` or `TEMP` environment variable for the `/tmp` value and maybe the `MSYSTEM_PREFIX` value (or something similar) to derive a value for the `/` path.

I don't want to optimize too early though. I'm hoping that user feedback might identify other situations where we might need to use cygpath (or not).

It may even be possible to pull in the actual c source code of cygpath and use it together with CGO for a fully equivalent solution, but this will complicate builds. Maybe one day someone will take a stab at implementing cygpath entirely in Go.

To answer your question more directly: For how this PR uses cygpath, I think we could get away with a very basic version of it implemented in Go using the heuristic method proposed above.

It is also worth mentioning that the `stdlib.sh` already has a reference to cygpath (in the `source_env` function) and uses it when it is available. With that in mind, if it will already be called for every `source_env` invocation on msys2, then the two calls made in this PR are not that much of an overhead.

Reply:
@zimbatm I've made a slight optimization: cygpath is now only called once at direnv startup, specifically to resolve the PATH variable. Maybe we can leave the "implement cygpath logic in Go" for another PR?